File: java/src/main/java/org/wikimedia/metrics_platform/curation/rules/ComparableCurationRules.java
package org.wikimedia.metrics_platform.curation.rules;

import java.util.Collection;

import com.google.gson.annotations.SerializedName;

import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@Builder @EqualsAndHashCode @ToString
public class ComparableCurationRules<T extends Comparable<T>> {
    private T equals;
    @SerializedName("not_equals") private T notEquals;
    @SerializedName("greater_than") private T greaterThan;
    @SerializedName("less_than") private T lessThan;
    @SerializedName("greater_than_or_equals") private T greaterThanOrEquals;
    @SerializedName("less_than_or_equals") private T lessThanOrEquals;
    private Collection<T> in;
    @SerializedName("not_in") private Collection<T> notIn;

    @SuppressWarnings({"checkstyle:CyclomaticComplexity", "checkstyle:NPathComplexity"})
    public boolean apply(T value) {
        if (equals != null && !equals.equals(value)) {
            return false;
        }
        if (notEquals != null && notEquals.equals(value)) {
            return false;
        }
        if (greaterThan != null && greaterThan.compareTo(value) >= 0) {
            return false;
        }
        if (lessThan != null && lessThan.compareTo(value) <= 0) {
            return false;
        }
        if (greaterThanOrEquals != null && greaterThanOrEquals.compareTo(value) > 0) {
            return false;
        }
        if (lessThanOrEquals != null && lessThanOrEquals.compareTo(value) < 0) {
            return false;
        }
        if (in != null && !in.contains(value)) {
            return false;
        }
        if (notIn != null && notIn.contains(value)) {
            return false;
        }
        return true;
    }
}

File: swift/Sources/WikimediaMetricsPlatform/Curation/Rules/ComparableCurationRules.swift
import Foundation

/**
 * Curation rules to be used with context values whose types conform to Comparable.
 * This is used for context values with numeric types.
 */
struct ComparableCurationRules<T: Comparable & Hashable & Decodable>: Decodable {
    var equals: T?
    var notEquals: T?
    var greaterThan: T?
    var lessThan: T?
    var greaterThanOrEquals: T?
    var lessThanOrEquals: T?
    var inCollection: [T]?
    var notInCollection: [T]?

    init(
            equals: T? = nil,
            notEquals: T? = nil,
            greaterThan: T? = nil,
            lessThan: T? = nil,
            greaterThanOrEquals: T? = nil,
            lessThanOrEquals: T? = nil,
            inCollection: [T]? = nil,
            notInCollection: [T]? = nil
    ) {
        self.equals = equals
        self.notEquals = notEquals
        self.greaterThan = greaterThan
        self.lessThan = lessThan
        self.greaterThanOrEquals = greaterThanOrEquals
        self.lessThanOrEquals = lessThanOrEquals
        self.inCollection = inCollection
        self.notInCollection = notInCollection
    }

    func apply(to value: T) -> Bool {
        if equals != nil && value != equals! {
            return false
        }
        if notEquals != nil && value == notEquals! {
            return false
        }
        if greaterThan != nil && value <= greaterThan! {
            return false
        }
        if lessThan != nil && value >= lessThan! {
            return false
        }
        if greaterThanOrEquals != nil && value < greaterThanOrEquals! {
            return false
        }
        if lessThanOrEquals != nil && value > lessThanOrEquals! {
            return false
        }
        if inCollection != nil && !inCollection!.contains(value) {
            return false
        }
        if notInCollection != nil && notInCollection!.contains(value) {
            return false
        }
        return true
    }

    enum CodingKeys: String, CodingKey {
        case equals
        case notEquals = "not_equals"
        case greaterThan = "greater_than"
        case lessThan = "less_than"
        case greaterThanOrEquals = "greater_than_or_equals"
        case lessThanOrEquals = "less_than_or_equals"
        case inCollection = "in"
        case notInCollection = "not_in"
    }
}
